scene/resources/material.cpp

```diff
@@ -1404,6 +1409,21 @@ void fragment() {)";
 )";
 	}
 
+	if (depth_offset_enabled) {
+		code += R"(
+	// Depth Offset: Enabled
+	// Force transparency on the material (required for depth offset).
+	ALPHA = 1.0;
+	vec4 view_position = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2. - 1.0, FRAGCOORD.z, 1.0);
+	vec3 view_direction = normalize(VIEW);
+	view_position.xyz /= view_position.w;
+	view_position.xyz -= depth_offset * view_direction;
+	vec4 ndc_position = PROJECTION_MATRIX * vec4(view_position.xyz, 1.0);
+	ndc_position.xyz /= ndc_position.w;
```

Comment:
If you're curious, this part is the exact same as the `DEPTH` adjustment code featured in https://github.com/godotengine/godot/pull/97646.

Reply:
This looks suspiciously like it's dependent on Vulkan-style NDCs where the NDC.z is in the range `[0,1]`, and therefore won't work correctly in Compatibility? But I only know just enough to use these in typical cases in user shaders, not quite enough to suggest a proper correction here.

Reply:
Yes, it likely needs an `#ifdef` check depending on the current rendering method to alter the generated code.

Reply:
I think the Compatibility view position snippet is this, a variation of what is documented [here](https://docs.godotengine.org/en/latest/tutorials/shaders/advanced_postprocessing.html#depth-texture):
```
#if CURRENT_RENDERER == RENDERER_COMPATIBILITY
vec4 view_position = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2. - 1.0, FRAGCOORD.z * 2. - 1.0, 1.0);
#else
vec4 view_position = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2. - 1.0, FRAGCOORD.z, 1.0);
#endif
```
or
```
#if CURRENT_RENDERER == RENDERER_COMPATIBILITY
vec4 view_position = INV_PROJECTION_MATRIX * vec4(vec3(SCREEN_UV, FRAGCOORD.z) * 2. - 1.0, 1.0);
#else
vec4 view_position = INV_PROJECTION_MATRIX * vec4(SCREEN_UV * 2. - 1.0, FRAGCOORD.z, 1.0);
#endif
```
But you probably also need to convert NDC.z again when setting depth:
```
#if CURRENT_RENDERER == RENDERER_COMPATIBILITY
DEPTH = ndc_position.z * 0.5 + 0.5;
#else
DEPTH = ndc_position.z;
#endif
```
Again, I only have casual user shader knowledge of this, and my understanding of the underlying coordinate spaces is still fuzzy.

Since the material writes out a GDShader, not GLSL directly, we still have access to the preprocessor here, right?

Reply:
Thanks! I'll look into integrating this.

> Since the material writes out a GDShader, not GLSL directly, we still have access to the preprocessor here, right?

Yes :slightly_smiling_face:

Reply:
It seems preprocessor macros don't work within the BaseMaterial3D shader: https://github.com/godotengine/godot/pull/97646#issuecomment-2529909391

This is likely an engine bug unless I missed something obvious.